### PR TITLE
Intelligent S3 bucket migration support

### PR DESCRIPTION
**Pull request by ddvdozuki**
# Problem
Our migration system had a fundamental flaw. We could take an existing RDS snapshot and existing S3 buckets to launch a new instance, however that old instance still has a terraform state entry for those S3 buckets. What happens when that old unneeded instance is torn down? The answer is the customer data goes away. 

# Solution
The easiest solution is to simply perform state surgery on that old instance and remove the references to those buckets before deleting and while that works it's far from ideal. One forgotten move or misstep and all of our customer's data is gone forever. I believe this potential calamity is reason enough to invest in a safer replication strategy.

We now create new buckets for each environment regardless of migration state. If migrating, we will setup replication rules on the old buckets and fire off batch replication jobs to copy existing objects from the source buckets to the destination buckets leaving a clear safety separation between environments.

## Caveats
As you may have noticed this has some caveats:
1. S3 bucket replication configurations only apply to **new** objects created after the configuration is applied. To copy **existing** objects, an S3 batch job is required. 
2. We fire off an S3 batch job to copy existing objects but it's asynchronous. It's possible to make this a synchronous process by adding additional logic to the batch creation script but I opted for an async approach due to the unknown and potentially expansive load times some of our larger customer's migrations would take causing TF timeouts and all kinds of messy results. Safety over convenience. So a manual verification of batch job completion is required before the old stack is torn down.
3. Potentially the most annoying caveat and the one the irks me the most is the fact that we have now tied our new state to our old stack (a reversal of what we had before but with zero danger) by creating the replication configuration on the source buckets. That old stack can no longer be torn down fully while the replication configuration is in place (can't delete the bucket when there's still a replication config attached). So to remedy this you would need to re-run this new stack (migration target) with migration disabled so it deletes the replication config thereby freeing up that old stack for deletion. 

Unfortunately there's a bit of a one-two approach to this but I believe on the whole it's a net positive gain for data safety and reliability. 

## Additional Changes
1. To support configuring replication in this way we needed to upgrade to the latest version of the AWS provider. This came with it many refactored resources. The functionality of said resources stays almost the same but the syntax and module implementation slightly differs. 
2. Since we're upgrading the AWS provider we might as well update the other providers we use as well. 
3. Some minor refactors for cleanliness and semantics. Like moving all the data resources out of the `main.tf` into their respective files.
4. Fix replicated install script from trying to re-install on ephemeral environments. 

## QA
I have tested the following scenarios:
* Deploying min, bi, hooks via workstation and cloudformation and validating the app works
* Deploying v3.1 min & full via workstation and cloudformation then upgrading to this branch and verifying nothing bad happens in the TF apply and the app still works
* Deploying this branch, migrating to a new copy of this branch and validating s3 buckets are replicated

Deploying to <= v3.1 and migrating to a new copy of this branch is not supported. Update the source instance first before migration.

Some potential issues that cropped up:
* if we upgrade a customer from <=v3.1 to this version then the ~database passwords~, self-signed certificates, and s3 KMS key will be changed. 
  * ~This requires a full reboot of the application with some minor downtime (<5 minutes) to load the new db credentials.~ (Fixed in 2fb7b5de8fbb07dae000ac13d025f047cab79dc2)
  * ~The bastion instance will need to be recycled to load the new db credentials.~
  * If this new upgraded stack were to ever be migrated in the future someone would need to run a batch job to normalize all the objects in the buckets to use the same key since any objects created before this update will be using the default KMS key and any after will be using the bespoke key.

### CODE REVIEW

File: terraform/physical/s3.tf [hunk 1]
```diff
@@ -1,252 +1,403 @@
...
+      sse_algorithm     = "aws:kms"
+    }
   }
 }
-resource "aws_s3_bucket_public_access_block" "guide_documents_acl_block" {
-  count = var.create_s3_buckets ? 1 : 0
 
-  bucket = aws_s3_bucket.guide_documents[0].id
+# Begin creating buckets and associated bucket resources dynamically
+resource "aws_s3_bucket" "guide_buckets" {
```
[the remark this answers is not in the document]
Comment (ddvdozuki) [reply terraform/physical/s3.tf:335]: tfsec is unable to parse the dynamic bucket creation and tell that bucket logging is indeed enabled

[the remark this answers is not in the document]
Comment (ddvdozuki) [reply terraform/physical/s3.tf:335]: tfsec is unable to parse the dynamic bucket creation and tell that bucket logging is indeed enabled

[the remark this answers is not in the document]
Comment (ddvdozuki) [reply terraform/physical/s3.tf:335]: tfsec is unable to parse the dynamic bucket creation and tell that bucket logging is indeed enabled

[the remark this answers is not in the document]
Comment (ddvdozuki) [reply terraform/physical/s3.tf:335]: tfsec is unable to parse the dynamic bucket creation and tell that bucket logging is indeed enabled

File: terraform/physical/s3.tf [hunk 2]
```diff
@@ -1,252 +1,403 @@
...
+data "aws_kms_key" "s3" {
+  count = var.s3_kms_key_id != "" ? 1 : 0
 
-  block_public_acls       = true
-  block_public_policy     = true
-  ignore_public_acls      = true
-  restrict_public_buckets = true
+  key_id = var.s3_kms_key_id
+}
+resource "aws_kms_key" "s3_kms_key" {
```
[the remark this answers is not in the document]
Comment (ddvdozuki) [reply terraform/physical/s3.tf:61]: I am not sure what the implication of enabling key rotation is so I did not do it by default.

**Review by scbarber — APPROVED** [review]
LGTM

**Comment by segrettodozuki** [comment]
I am not going to leave review that would block this from merging; however, I want to make sure we are in agreement on general principles regarding PRs.  My comments here are based on some assumptions.  Let me know how/where I am off.  We strive to keep our Jira stories small and concise.  It follows that we try to minimize 'scope creep' so we minimize the number of changes introduced in our PRs so that our PRs are focused on the completion of our story.  That said, this is a grey area and general small code cleanup is acceptable.  However, if a related agenda or dependent agenda is found when working a story, rather than complete it in the same PR and story as scope creep, it is better to create a second, possibly blocking dependent Jira story.  The blocking Story is then completed with a PR focused on the dependency.
For example, you mentioned, "Since we're upgrading the AWS provider we might as well update the other providers we use as well."  It is possible that this could have been its own Jira story and PR, focused on upgrading providers, which blocked the original Jira Story regarding allowing terraform to destroy an old migrated stack without removing customer S3 data in the new stack.  In this way the PRs would be more clearly focused for the reviewer.  With a separated Jira story and PR focused on upgrading providers I would agree it makes sense to do them all; however, when it is scope creep on an existing PR, it becomes more arguable that you shouldn't give into the mindset that, well because I need to do this I mine as well also do that, because that mindset leads you down a slippery slope that leads to more common scope creep in future tickets.
Increasing PR size increases risk to merging the PR, and it increases difficulty on the PR reviewer on catching all nuances a PR is delivering. There are more reasons for breaking up Jira Stories and PRs.  While it is clear we are not an agile team, nor are we completing scrum, some agile principles should still be applicable to us.  In this light, the textbook answer as to why stories should be small I think is applicable "increased focus, which helps prevent failure; earlier discovery / faster feedback; shorter lead time/ better throughput; and reduced testing overhead."  
I thought this was worth diving deep because I have seen a lot of really huge PRs; however, this was the first of them that was clear to me it could have potentially been broken up into smaller PRs and Jira stories.

File: terraform/physical/util/create-s3-batch.sh [hunk 3]
```diff
@@ -5,12 +5,18 @@ set -euo pipefail
 AWS_LOGGING_BUCKET="$1"
 AWS_SOURCE_BUCKET="$2"
 AWS_REPLICATION_ROLE="$3"
-AWS_PROFILE="$4"
-AWS_ACCOUNT="$5"
+AWS_ACCOUNT="$4"
+AWS_PROFILE="$5"
```
Comment (caphene) [inline terraform/physical/util/create-s3-batch.sh]: This isn't a problem, more of a curiosity: Why the position change here?

**Review by caphene — APPROVED** [review]
I left one question as a curiosity. Otherwise everything looks good. 👍

File: terraform/physical/util/create-s3-batch.sh (hunk 3 above)
Comment (ddvdozuki) [reply terraform/physical/util/create-s3-batch.sh]: Because the profile selection is optional so it needs to go last.

**Pull request merged**